### PR TITLE
Fix: Allow Nextest to Work with Manifests Outside of Workspace Root Path (nextest-rs/nextest issue #1684)

### PR DESCRIPTION
**Pull request by PegasusPlusUS**
Fix: Allow Nextest to Work with Manifests Outside of Workspace Root Path
Description
This pull request addresses an issue where Nextest fails to function correctly when the Cargo.toml manifest file is located outside of the workspace root path. The changes ensure that Nextest can handle such scenarios gracefully.

Changes
Updated the guppy crate to include a helper function for resolving paths outside the workspace root.

Modified the guppy/src/graph/build.rs fn workspace_path() #402~#410 to incorporate the new path resolution logic in case there's out of root.
Helper fn find_relative_path_utf8() placed at the end of build.rs, together with test case.

Issue
This pull request fixes nextest-rs/nextest issue #1684.

Testing
Verified that Nextest can now correctly locate and use Cargo.toml files outside the workspace root.

Ran existing test suites to ensure no regressions were introduced.

### CODE REVIEW

**Comment by PegasusPlusUS** [comment]
Commit squash refactor

File: guppy/src/graph/build.rs [hunk 1]
```diff
@@ -1002,6 +1008,42 @@ fn convert_forward_slashes<'a>(rel_path: impl Into<Cow<'a, Utf8Path>>) -> Utf8Pa
     }
 }
 
+// Calculate the relative path from `from` to `to`.
+// This function finds the relative path between two given paths.
+// It first identifies the common prefix between the two paths and then
+// constructs the relative path by adding ".." for each remaining component
+// in the `from` path and appending the remaining components from the `to` path.
+#[track_caller]
+pub fn find_relative_path_utf8(from: &Utf8Path, to: &Utf8Path) -> Utf8PathBuf {
```
Comment (sunshowers) [inline guppy/src/graph/build.rs:1017]: This function should live on `PackageSourceImpl`, next to `PackageSourceImpl::create_path` which does something similar for path dependencies.

**Review by sunshowers — COMMENTED** [review]
Thanks for working on this! I have a few comments.

**Comment by PegasusPlusUS** [comment]
OK, will continue update here

File: guppy/src/graph/build.rs [hunk 2]
```diff
@@ -399,15 +399,21 @@ impl<'a> GraphBuildState<'a> {
...
+        let workspace_path = if let Ok(stripped_workspace_path) =
+            manifest_path.strip_prefix(self.workspace_root)
+        {
+            stripped_workspace_path
+        } else {
+            // Error::PackageGraphConstructError(format!(
+            //     "workspace member '{}' at path {} not in workspace (root: {})",
+            //     id, manifest_path, self.workspace_root
+            // ));
+            // If manifest path is out of workspace root, try find relative path instead
```
[the remark this answers is not in the document]
Comment (PegasusPlusUS) [reply guppy/src/graph/build.rs:413]: ok

File: guppy/src/graph/build.rs [hunk 3]
```diff
@@ -399,15 +399,21 @@ impl<'a> GraphBuildState<'a> {
...
-            .map_err(|_| {
-                Error::PackageGraphConstructError(format!(
-                    "workspace member '{}' at path {} not in workspace (root: {})",
-                    id, manifest_path, self.workspace_root
-                ))
-            })?;
+        // Try to strip off the workspace path from the manifest path.
+        let _utf8_path_buf; // relative path lifetime helper
+        let workspace_path = if let Ok(stripped_workspace_path) =
+            manifest_path.strip_prefix(self.workspace_root)
```
[the remark this answers is not in the document]
Comment (PegasusPlusUS) [reply guppy/src/graph/build.rs:405]: OK

File: guppy/src/graph/build.rs [hunk 4]
```diff
@@ -1002,6 +1008,42 @@ fn convert_forward_slashes<'a>(rel_path: impl Into<Cow<'a, Utf8Path>>) -> Utf8Pa
     }
 }
 
+// Calculate the relative path from `from` to `to`.
+// This function finds the relative path between two given paths.
+// It first identifies the common prefix between the two paths and then
+// constructs the relative path by adding ".." for each remaining component
+// in the `from` path and appending the remaining components from the `to` path.
```
[the remark this answers is not in the document]
Comment (PegasusPlusUS) [reply guppy/src/graph/build.rs:1015]: OK, using pathdiff, revised update will only modify one line of original code, and add 3 test cases for test different paths situations. As test, Cargo test support manifest on other driver, to updated guppy will behavior the same way.

File: guppy/src/graph/build.rs [hunk 5]
```diff
@@ -1061,4 +1103,18 @@ mod tests {
         // This should have forward slashes, even on Windows.
         assert_eq!(path.as_str(), "../../foo/bar/baz.txt");
     }
+
+    #[test]
+    fn test_workspace_path_out_of_pocket() {
```
[the remark this answers is not in the document]
Comment (PegasusPlusUS) [reply guppy/src/graph/build.rs:1108]: Add 2 extra test cases for windows,

**Comment by PegasusPlusUS** [comment]
Poor, found huge amount of background file commits, (I'm using GH codespace), made the PR quite confusing, :(. Wish I can close this and initialize a clean new one...

**Comment by PegasusPlusUS** [comment]
Will close this and resubmitt (after test on windows for demo repo) because it has been polluted by a background task commit at GitHub codespace

**Comment by sunshowers** [comment]
Sounds fine — sorry that codespaces have been such trouble!

**Comment by PegasusPlusUS** [comment]
Resubmit at PR #331

**Pull request closed without merging**